### PR TITLE
implemented config reconfiguration

### PR DESCRIPTION
**Pull request by cryogenian**

### CODE REVIEW

**Comment by alissapajer** [comment]
[ch11361]

File: datasource/src/test/scala/quasar/physical/s3/S3DatasourceModuleSpec.scala [hunk 1]
```diff
@@ -109,6 +110,93 @@ class S3DatasourceModuleSpec extends Specification {
...
+        "region" := Json.jString("r")))
+
+    "returns malformed error if patch or source can't be decoded" >> {
+      val incorrect = Json()
+
+      "both" >> {
+        S3DatasourceModule.reconfigure(incorrect, incorrect) must beLeft(
+          MalformedConfiguration(
+            S3DatasourceModule.kind,
+            S3DatasourceModule.sanitizeConfig(incorrect),
```
Comment (rintcius) [inline datasource/src/test/scala/quasar/physical/s3/S3DatasourceModuleSpec.scala:139]: can use `incorrect` without `S3DatasourceModule.sanitizeConfig`

File: datasource/src/test/scala/quasar/physical/s3/S3DatasourceModuleSpec.scala [hunk 2]
```diff
@@ -109,6 +110,93 @@ class S3DatasourceModuleSpec extends Specification {
...
+            "Source configuration in reconfiguration is malformed."))
+      }
+      "source" >> {
+        S3DatasourceModule.reconfigure(incorrect, patchJson) must beLeft(
+          MalformedConfiguration(
+            S3DatasourceModule.kind,
+            S3DatasourceModule.sanitizeConfig(incorrect),
```
Comment (rintcius) [inline datasource/src/test/scala/quasar/physical/s3/S3DatasourceModuleSpec.scala:146]: same

File: datasource/src/test/scala/quasar/physical/s3/S3DatasourceModuleSpec.scala [hunk 3]
```diff
@@ -109,6 +110,93 @@ class S3DatasourceModuleSpec extends Specification {
...
+            "Source configuration in reconfiguration is malformed."))
+      }
+      "patch" >> {
+        S3DatasourceModule.reconfigure(sourceJson, incorrect) must beLeft(
+          MalformedConfiguration(
+            S3DatasourceModule.kind,
+            S3DatasourceModule.sanitizeConfig(incorrect),
```
Comment (rintcius) [inline datasource/src/test/scala/quasar/physical/s3/S3DatasourceModuleSpec.scala:153]: same

File: datasource/src/main/scala/quasar/physical/s3/S3DatasourceModule.scala [hunk 4]
```diff
@@ -84,14 +79,33 @@ object S3DatasourceModule extends LightweightDatasourceModule {
...
+          "Source configuration in reconfiguration is malformed."))
+
+    patch <- patchJson.as[S3Config].result.leftMap(_ =>
+      DatasourceError
+        .MalformedConfiguration[Json](
+          kind,
+          sanitizeConfig(patchJson),
+          "Patch configuration in reconfiguration is malformed."))
+
+    reconfigured <- original.reconfigure(patch).leftMap(c =>
```
Comment (rintcius) [inline datasource/src/main/scala/quasar/physical/s3/S3DatasourceModule.scala:97]: 👍 That's nicer than what I have - I'll update my PR too

**Comment by cryogenian** [comment]
@rintcius Addressed

**Comment by cryogenian** [comment]
Random test noizes 😬 restarted and it worked

**Pull request merged**